object-cache.php

```diff
@@ -1000,46 +1044,60 @@ protected function _connect_redis() {
 			}
 		}
 
-		$this->redis = new Redis;
-
 		if ( file_exists( $redis_server['host'] ) && 'socket' === filetype( $redis_server['host'] ) ) { //unix socket connection
 			//port must be null or socket won't connect
 			$port = null;
 		} else { //tcp connection
 			$port = ! empty( $redis_server['port'] ) ? $redis_server['port'] : 6379;
 		}
-		$this->redis->connect( $redis_server['host'], $port, 1, null, 100 ); // 1s timeout, 100ms delay between reconnections
-		$keys_methods = array(
-			'auth'     => 'auth',
-			'database' => 'select',
+
+		$defaults = array(
+			'host' => $redis_server['host'],
+			'port' => $port,
+			'timeout' => 1000, // TODO I multiplied this by 1000 so we'd have a common measure of ms instead of s and ms, need to make sure this gets divided by 1000
+			'retry_interval' => 100,
 		);
+		// 1s timeout, 100ms delay between reconnections
+
+		// merging the defaults with the original $redis_server enables any
+		// custom parameters to get sent downstream to the redis client.
+		return array_replace_recursive( $redis_server, $defaults );
+	}
+
+	public function client_connection( $connection_details ) {
+		$redis = new Redis;
+
+		// TODO should we cache this connection?
+		$redis->connect(
+			$connection_details['host'],
+			$connection_details['port'],
+			// $connection_details['timeout'] is sent in milliseconds,
+			// connect() takes seconds, so divide by 1000
+			$connection_details['timeout'] / 1000,
+			null,
+			$connection_details['retry_interval']
+		);
+
+		return $redis;
+	}
+
+	function setup_client_connection( $redis, $settings, $keys_methods ) {
 		foreach ( $keys_methods as $k => $method ) {
-			if ( ! isset( $redis_server[ $k ] ) ) {
+			if ( ! isset( $settings[ $k ] ) ) {
 				continue;
 			}
 			try {
-				$this->redis->$method( $redis_server[ $k ] );
+				$redis->$method( $settings[ $k ] );
 			} catch ( RedisException $e ) {
+
 				// PhpRedis throws an Exception when it fails a server call.
 				// To prevent WordPress from fataling, we catch the Exception.
-				try {
-					$this->last_triggered_error = 'WP Redis: ' . $e->getMessage();
-					// Be friendly to developers debugging production servers by triggering an error
-					// @codingStandardsIgnoreStart
-					trigger_error( $this->last_triggered_error, E_USER_WARNING );
-					// @codingStandardsIgnoreEnd
-				} catch ( PHPUnit_Framework_Error_Warning $e ) {
-					// PHPUnit throws an Exception when `trigger_error()` is called.
-					// To ensure our tests (which expect Exceptions to be caught) continue to run,
-					// we catch the PHPUnit exception and inspect the RedisException message
-				}
+				// TODO Perhaps we catch and rethrow? or return instance of
+				// WP_Error?
```

Comment:
I also just realized I missed some functionality that needs to update in `_call_redis` because it's trying to catch `RedisExceptions`. The direction I take with that sort of depends on your answer to the above.